### PR TITLE
IGNITE-21348 Trigger the lease negotiation retry in case when the lease candidate is no more contained in assignments

### PR DESCRIPTION
**Pull request by denis-chudov**
https://issues.apache.org/jira/browse/IGNITE-21348

### CODE REVIEW

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java [hunk 1]
```diff
@@ -368,15 +369,14 @@ private void updateLeaseBatchInternal() {
                         writeNewLease(grpId, lease, candidate, renewedLeases, toBeNegotiated);
 
                         continue;
+                    } else {
+                        agreement.checkValid(grpId, topologyTracker.currentTopologySnapshot(), assignments);
```
Comment (vldpyatkov) [inline modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java:373]: I would put it up just after getting an agreement with the additional conditions.
It makes logic simpler.

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/negotiation/LeaseNegotiator.java [hunk 2]
```diff
@@ -88,28 +91,37 @@ public void negotiate(Lease lease, boolean force) {
...
      *
      * @param groupId Replication group id.
      * @return Lease agreement.
      */
-    public LeaseAgreement negotiated(ReplicationGroupId groupId) {
-        LeaseAgreement agreement = leaseToNegotiate.getOrDefault(groupId, UNDEFINED_AGREEMENT);
+    public LeaseAgreement getAndRemoveIfReady(ReplicationGroupId groupId) {
+        LeaseAgreement[] res = new LeaseAgreement[1];
+
+        leaseToNegotiate.compute(groupId, (k, v) -> {
```
Comment (vldpyatkov) [inline modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/negotiation/LeaseNegotiator.java:118]: If a race is excluded here, I prefer to use plain read and conditinal remove.

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/negotiation/LeaseNegotiator.java [hunk 3]
```diff
@@ -42,7 +41,11 @@ public class LeaseNegotiator {
 
     private static final PlacementDriverMessagesFactory PLACEMENT_DRIVER_MESSAGES_FACTORY = new PlacementDriverMessagesFactory();
 
-    /** Leases ready to accept. */
+    static LeaseGrantedMessageResponse NOT_ACCEPTED_RESPONSE = PLACEMENT_DRIVER_MESSAGES_FACTORY.leaseGrantedMessageResponse()
```
Comment (vldpyatkov) [inline modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/negotiation/LeaseNegotiator.java:44]: Why don't you use {@code null}, like it is in the undefined agreement?

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/TopologyTracker.java [hunk 4]
```diff
@@ -142,15 +146,6 @@ public void onUpdate(LogicalTopologySnapshot topologySnap) {
...
             LOG.debug("Logical topology updated for placement driver [topologySnap={}]", topologySnap);
-
-            triggerToRenewLeases();
         }
     }
-
-    /**
-     * Triggers to renew leases forcibly. The method wakes up the monitor of {@link LeaseUpdater}.
-     */
-    private void triggerToRenewLeases() {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/TopologyTracker.java:153]: why is it still needed?

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java (hunk 1 above)
Comment (denis-chudov) [reply modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java:373]: we don't need to check the validity if the agreement is ready (accepted also means ready), that's why I put it to `else` block

File: modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/negotiation/LeaseNegotiator.java (hunk 2 above)
Comment (denis-chudov) [reply modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/negotiation/LeaseNegotiator.java:118]: I wouldn't say so, this is a concurrent map which can be changed from different threads (lease updater and `whenComplete` of MS invoke

**Pull request merged**